File: models/symbolic_solvers/Prover9/ladr/kol.pl
#!/usr/bin/perl
use strict;
use warnings;

#print "Hello, world!\n";

my $MIN = 2;

open(INFILE, "/dev/stdin") or die "Can't open the input file...\n";
while (<INFILE>) {
#   print "% Here's the line: $_";
  my $total = 0;
  my $string = $_; # foobarfoobar
  my $length = (length ($string)) - 2;
  my $index = $MIN;
#  print "Will check from index $index to $length.\n";
#  print "Length of string is $length.\n";
  while($index <= $length) { # index of where the search begins
     my $maxlook = ($length - $index) + 2;
     if ($maxlook > $index) { $maxlook = $index; }# don't check for occurences longer than head of string
#     print "Checking at index: $index.\n";
#     print "Will check strings of length $MIN through $maxlook.\n";
     my $flen = 0; # length of found redundancy
     for(my $window = $MIN; $window <= $maxlook; $window++) { # index of length of lookahead to check
        my $tocheck = substr( $string, $index, $window );
#         print "Want to know if $tocheck occurs in positions 0 until $index.\n";
        my $ind = index($string, $tocheck, 0);
        if( $ind < $index ) {$flen = $window; }
     } # end lookahead buffer loop
     if($flen == 0) { $index++; }
     else           { $index = $index + $flen; $total = $total + $flen }
#      print "total: $total\n";
#      print "index: $index\n";
#      print "length: $length\n";
#      print "MIN: $MIN\n";
  } # end outer while loop
#   print "DONE.\n";
  my $kolmogorov = ($total / ($length-1));
#   print "TOTAL REDUNDANCY: $total\n";
  print "$kolmogorov\n"

} # end of while loop

close(INFILE);

File: models/symbolic_solvers/Prover9/ladr/kol.py
#!/usr/bin/python

import sys

# print "Hello, world!"

MIN = 2

def kolmogorov(string):
   print "string=%s" % string
   total = 0
   length = len(string) - 2
   index = MIN
   print "Will check from index %d to %d." % (index,length)
   print "Length of string is %d." % length
   while index <= length:  # index of where the search begins
       maxlook = (length - index) + 2
       # don't check for occurences longer than head of string
       if maxlook > index:
           maxlook = index
       print "Checking at index: %d." % index
       print "Will check strings of length %d through %d." % (MIN,maxlook)
       flen = 0 # length of found redundancy
       # index of length of lookahead to check
       for window in range(MIN, maxlook+1):  # note: range(2,5) is [2,3,4]
           tocheck = string[ index : index+window ]  # note: "abcde"[1:3] is "bc"
           print "Want to know if %s occurs in pos 0 until %d." % (tocheck,index)
           ind = string.find(tocheck)
           if ind < index:
               flen = window

       if flen == 0:
           index += 1
       else:
           index += flen
           total += flen
       print "total=%d, index=%d, length=%d, MIN=%d" % (total,index,length,MIN)
   print "DONE.";
   print "TOTAL REDUNDANCY: %d" % total
   return float(total) / (length-1)

while 1:
   string = sys.stdin.readline()
   if not string:
       break;
   print str(kolmogorov(string))

